src/more-optimizations/cached-dataset.md

```diff
@@ -0,0 +1,145 @@
+
+## Usage
+
+[`CachedDataset`](https://bagua.readthedocs.io/en/latest/autoapi/bagua/torch_api/contrib/index.html#bagua.torch_api.contrib.CachedDataset) is
+a Pytorch custom dataset (see [Creating a Custom Dataset for your files](https://pytorch.org/tutorials/beginner/basics/data_tutorial.html#creating-a-custom-dataset-for-your-files)).
+It wraps a Pytorch dataset and caches its samples into a distributed key-value store. We can specify the backend to
+use on the initialization of a cached dataset. Currently [Redis](https://redis.io/) is supported, which is an **in-memory** data store.
+
+By default, cached dataset will spawn a new Redis instance on each worker node, and data is sharded across all
+Redis instances on all nodes in the Bagua job. We can specify the maximum memory limit to use for each node, by passing
+`capacity_per_node` to `CachedDataset`.
+
+The following is an example to use a Redis-backend cached dataset, the maximum memory limit on each node is `400GB`. A
+4-node Bagua job can have a maximum memory limit of `1.6TB`.
+
+```python
+from bagua.torch_api.contrib import CachedDataset
+
+cache_dataset = CachedDataset(
+ dataset,
+ backend="redis",
+ dataset_name="ds",
+ capacity_per_node=400 * 1024 * 1024 * 1024,
+)
+dataloader = torch.utils.data.DataLoader(cached_dataset)
+
+for i, (input, target) in enumerate(dataloader):
+    ...
+```
+
+By setting `cluster_mode=False`, we can restrict each training node to use only its local Redis instance.
+
+```python
+cache_dataset = CachedDataset(
+ dataset,
+ backend="redis",
+ dataset_name="ds",
+ cluster_mode=False,
+ capacity_per_node=400 * 1024 * 1024 * 1024,
+)
+```
+
+We can also use existing Redis servers as the backend store by passing a list of host information of redis servers to `hosts`.
+
+```python
+hosts = [
+    {"host": "192.168.1.0", "port": "7000"},
+    {"host": "192.168.1.1", "port": "7000"},
+]
+cache_dataset = CachedDataset(
+    dataset,
+    backend="redis",
+    dataset_name="ds",
+    hosts=hosts,
+)
+```
+
+### Multiple cached dataset
+
+Multiple cached dataset shares the same backend store, thus we need to specify a unique name for each dataset to avoid
+overwriting samples from each other.
+
+```python
+from bagua.torch_api.contrib import CachedDataset
+
+dataset1 = ...
+dataset2 = ...
+
+cache_dataset1 = CachedDataset(
+ dataset1,
+ backend="redis",
+ dataset_name="ds1",
+ capacity_per_node=400 * 1024 * 1024 * 1024,
+)
+
+cache_dataset1 = CachedDataset(
+ dataset2,
+ backend="redis",
+ dataset_name="ds2",
+ capacity_per_node=400 * 1024 * 1024 * 1024,
+)
+```
+
+It should be noted that the maximum memory for on each node is `400GB`. Cached dataset will reuse the same Redis instance
+on each node. Only parameters to spawn the first Redis instance will take effect.
```

Comment:
`Cached dataset will reuse the same Redis instance
on each node. Only parameters to spawn the first Redis instance will take effect.`
Not clear what this means - does the `ds2` in this example not take effect?

Reply:
Yes. `ds2` will take effect. 

`cluster_mode` and `capacity_per_node` are the parameters used to spawn redis instance. see https://bagua.readthedocs.io/en/latest/autoapi/bagua/torch_api/contrib/utils/redis_store/index.html#bagua.torch_api.contrib.utils.redis_store.RedisStore.  When we spawn a redis instance, we create a new instance of `RedisStore`. Only when `hosts=None`, we will spawn a new instance, otherwise we will use existing instances passed by `hosts`.

Reply:
Ok, that is much clearer now. Thanks!